dataflow/src/ops/rewrite.rs

```diff
@@ -0,0 +1,240 @@
+        }
+    }
+}
+
+
+impl Ingredient for Rewrite {
+    fn take(&mut self) -> NodeOperator {
+        Clone::clone(self).into()
+    }
+
+    fn ancestors(&self) -> Vec<NodeIndex> {
+        vec![self.src.as_global(), self.should_rewrite.as_global()]
+    }
+
+    fn is_join(&self) -> bool {
+        true
+    }
+
+
+    fn must_replay_among(&self) -> Option<HashSet<NodeIndex>> {
+        Some(Some(self.src.as_global()).into_iter().collect())
+    }
+
+    fn on_connected(&mut self, _: &Graph) {}
+
+    fn on_commit(&mut self, _: NodeIndex, remap: &HashMap<NodeIndex, IndexPair>) {
+        self.src.remap(remap);
+        self.should_rewrite.remap(remap);
+    }
+
+    fn on_input(
+        &mut self,
+        from: LocalNodeIndex,
+        rs: Records,
+        _: &mut Tracer,
+        _: Option<usize>,
+        nodes: &DomainNodes,
+        state: &StateMap,
+    ) -> ProcessingResult {
+        assert!(from == *self.src || from == *self.should_rewrite);
+        let misses = Vec::new();
+        let mut emit_rs = Vec::with_capacity(rs.len());
+
+        if rs.is_empty() {
+            return ProcessingResult {
+                results: rs,
+                misses: vec![],
+            };
+        }
+
+        for r in rs {
+            if from == *self.src {
+                let key = r[self.key].clone();
+                // ask should_rewrite if column should be rewritten
+                let rc = self.lookup(
+                    *self.should_rewrite,
+                    &[0],
+                    &KeyType::Single(&key),
+                    nodes,
+                    state,
+                ).unwrap();
+
+
+                if rc.is_none() {
+                    // todo (larat): might need to do some stuff here
+                    continue;
+                }
+
+                let mut rc = rc.unwrap().peekable();
+                if rc.peek().is_none() {
+                    emit_rs.push(r);
+                } else {
+                    let (r, positive) = r.extract();
+                    let row = self.generate_row(self.rewrite(r), positive);
+                    emit_rs.push(row);
+                }
+
+            } else if from == *self.should_rewrite {
+                let key = r[0].clone();
+                let other_rows = self.lookup(
+                    *self.src,
+                    &[self.key],
+                    &KeyType::Single(&key),
+                    nodes,
+                    state,
+                ).unwrap();
+
+                if other_rows.is_none() {
+                    // todo (larat): might need to do some stuff here
+                    continue;
+                }
+
+                let mut other_rows = other_rows.unwrap().peekable();
+                if other_rows.peek().is_none() {
+                    continue;
+                } else {
+                    let mut other;;
+                    while other_rows.peek().is_some() {
+                        other = other_rows.next().unwrap();
```

Comment:
Given that you always use `other.to_vec` below, I'd just do:
```rust
let other = other_rows.next().unwrap().to_vec();
```
And then `clone()` the thing you pass to the first call to `generate_row` below (or `.into()` if you switch to using that).